test/test_grid.py

```diff
@@ -299,3 +293,34 @@ def test_populate_lonlat_coord(self):
             nt.assert_almost_equal(vgrid.ds["Mesh2_node_y"].values[i],
                                    lat_deg[i],
                                    decimal=12)
+
+
+class TestConnectivity(TestCase):
+    ugrid_filepath = current_path / "meshfiles" / "ugrid" / "outRLL1deg" / "outRLL1deg.ug"
+    mpas_filepath = current_path / "meshfiles" / "mpas" / "QU" / "mesh.QU.1920km.151026.nc"
+    exodus_filepath = current_path / "meshfiles" / "exodus" / "outCSne8" / "outCSne8.g"
+
+    def test_build_edge_nodes(self):
```

Comment:
@philipc2 
Can you add more test cases here to make sure it works?
Especially the test cases that include different face geometry (The input Grid that contains the fill_values)

You can refer to my initial commit for this PR, which contains 3 input files as test cases for `test_build_face_edges_connectivity`

Reply:
Added a separate test case for Euler's formula using the datasets you mentioned.

All seems well, except one of the datasets that I tested `ugrid/geoflow-small/grid.nc` is not adhering to euler's formula. Will investigate this, but hoping it's not an issue with our implementation.

Reply:
Additionally, `exodus/mixed/mixed.nc` is also failing.

Reply:
All tests seem to be passing now; so did you fix them?

Reply:
> 
Since this implementation is the same as my old ones, the implementation itself should be all right. I have tested on mixed geometry faces and it still adheres the Euler's formula. 

There might be little bugs inside the codes that cause the inconsistency. Have you checked that all mixed faces are correctly patched and identified?

Reply:
> 
Since this implementation is the same as my old ones, the implementation itself should be all right. I have tested on mixed geometry faces and it still adheres the Euler's formula. 

There might be little bugs inside the codes that cause the inconsistency. Have you checked that all mixed faces are correctly patched and identified?

Reply:
@erogluorhan 

I've been locally testing the two failing datasets. I'm trying to figure out why those two specific datasets don't adhere to Euler's formula, especially since the geoflow dataset does not contain any fill values.

Reply:
@philipc2 

I digged into the codes a little bit for `geoflow-small/grid.nc`, 

I noticed that some of the edges are used only once in the face, indicating that there might be holes in the grid.

The Euler Formular in the form of `V-E+F=2.` is only valid for convex polyhedron [Euler](https://en.wikipedia.org/wiki/Euler_characteristic)

In our cases, it means that the Grids should cover the entire sphere (A.K.A. every edge will be used twice, A.K.A. there's no hole on the sphere)

Can you check if this `geoflow-small/grid.nc` has a hole or not? If there is, we don't need to worry about here, if there's not, we will dig more into what happened. Thanks!

Reply:
> @philipc2
> 
> I digged into the codes a little bit for `geoflow-small/grid.nc`,
> 
> I noticed that some of the edges are used only once in the face, indicating that there might be holes in the grid.
> 
> The Euler Formular in the form of `V-E+F=2.` is only valid for convex polyhedron ([Euler Formula](https://en.wikipedia.org/wiki/Euler_characteristic))
> 
> In our cases, it means that the Grids should cover the entire sphere (A.K.A. every edge will be used twice, A.K.A. there's no hole on the sphere)
> 
> Can you check if this `geoflow-small/grid.nc` has a hole or not? If there is, we don't need to worry about here, if there's not, we will dig more into what happened. Thanks!

This is a great catch, let me look into it.

Reply:
It doesn't seem like there is any holes in `geoflow-small/grid.n`, however this is the dataset I used last summer, and it does have multiple faces that cross the antemeridian. It uses positive 180 lon regardless of whether the face is on the left hand side (< 0 lon) or right hand side (>0 lon). Could this be part of the issue?

Reply:
> It doesn't seem like there is any holes in `geoflow-small/grid.n`, however this is the dataset I used last summer, and it does have multiple faces that cross the antemeridian. It uses positive 180 lon regardless of whether the face is on the left hand side (< 0 lon) or right hand side (>0 lon). Could this be part of the issue?

What I found for this dataset is that some unique edges are only used once, which should not happen if this dataset doesn't have any holes inside.

According to your explanation, it might be caused by the inconsistent longitude representation. Can you elaborate what does " It uses positive 180 lon regardless of whether the face is on the left hand side (< 0 lon) or right hand side (>0 lon)." mean here?

All my algorithms will base on the fact that all vertexes are represented as `[lon_degree, lat_degree]`, where `0<= lon_degree<=360`(I should take care of the 360 degree cases in my latlonbound function), and `-90<= lat_degree<=90`

Reply:
> What I found for this dataset is that some unique edges are only used once, which should not happen if this dataset doesn't have any holes inside. 

From my testing, it appears that each edge is used at least twice. Below is what I used to test it:
```Python
        grid_xr = xr.open_dataset(self.ugrid_filepath_04)
        grid_ux = ux.Grid(grid_xr)
        edge_nodes = grid_ux.Mesh2_edge_nodes.values.ravel()
        _, counts = np.unique(edge_nodes, return_counts=True)
```

> Can you elaborate what does " It uses positive 180 lon regardless of whether the face is on the left hand side (< 0 lon) or right hand side (>0 lon)." mean here?

Assuming our longitude values are [-180, 180] and the antimeridian is defined at +- 180, I'm saying that if a face is on the left hand plane (longitude < 0) and contains an edge that touches the antemeridian, it uses **positive** 180 instead of **negative** 180.

Reply:
I used a python dict to count the occurrence of each unique edge inside the `Grid._build_edge_node_connectivity` after line 427

```python
        # unique edge nodes
        edge_nodes_unique, inverse_indices = np.unique(edge_nodes,
                                                       return_inverse=True,
                                                       axis=0)

        # Use python dict to count the occurance of each face
        counts = {}
        for num in inverse_indices:
            if num in counts:
                counts[num] += 1
            else:
                counts[num] = 1
```
And here's what inside the `count`

```python
{0: 1, 3: 2, 9: 2, 1: 1, 2: 1, 5: 2, 11: 2, 4: 1, 7: 2, 13: 2, 6: 1, 8: 1, 15: 2, 12: 2, 18: 2, 10: 1, 14: 2, 20: 2, 16: 2, 22: 2, 17: 1, 24: 2, 21: 2, 27: 2, 19: 1, 23: 2, 29: 2, 25: 2, 31: 2, 26: 1, 33: 2, 30: 2, 36: 1, 28: 1, 32: 2, 37: 1, 34: 2, 38: 1, 35: 1, 39: 1, 40: 1, 43: 2, 49: 2, 41: 1, 42: 1, 45: 2, 51: 2, 44: 1, 47: 2, 53: 2, 46: 1, 48: 1, 55: 2, 52: 2, 58: 2, 50: 1, 54: 2, 60: 2, 56: 2, 62: 2, 57: 1, 64: 2, 61: 2, 67: 2, 59: 1, 63: 2, 69: 2, 65: 2, 71: 2, 66: 1, 73: 2, 70: 2, 76: 1, 68: 1, 72: 2, 77: 1, 74: 2, 78: 1, 75: 1, 79: 1, 80: 1, 83: 2, 89: 2, 81: 1, 82: 1, 85: 2, 91: 2, 84: 1, 87: 2, 93: 2, 86: 1, 88: 1, 95: 2, 92: 2, 98: 2, 90: 1, 94: 2, 100: 2, 96: 2, 102: 2...
```

Reply:
> > What I found for this dataset is that some unique edges are only used once, which should not happen if this dataset doesn't have any holes inside.
> 
> From my testing, it appears that each edge is used at least twice. Below is what I used to test it:
> 
> ```python
>         grid_xr = xr.open_dataset(self.ugrid_filepath_04)
>         grid_ux = ux.Grid(grid_xr)
>         edge_nodes = grid_ux.Mesh2_edge_nodes.values.ravel()
>         _, counts = np.unique(edge_nodes, return_counts=True)
> ```
> 
> > Can you elaborate what does " It uses positive 180 lon regardless of whether the face is on the left hand side (< 0 lon) or right hand side (>0 lon)." mean here?
> 
> Assuming our longitude values are [-180, 180] and the antimeridian is defined at +- 180, I'm saying that if a face is on the left hand plane (longitude < 0) and contains an edge that touches the antemeridian, it uses **positive** 180 instead of **negative** 180.

I tried your testing method. 

First, an edge should **NOT** shared by more than 2 faces.

Second, why do you want to use `ravel()` here? We want to count the occurrence of each edge, aka the occurrence of  each unique vertice pair

Reply:
@hongyuchen1030 

I will revisit this PR this weekend / early next week since it is a busy week for me.

Reply:
Thank you!

I haven't been able to find anything wrong with this dataset, but on an unrelated note it has also broken something with the redesign. I think it's fair to leave it out of testing until we can figure out more information about it.

Considering the other test cases, especially the one that compares it to the MPAS edge nodes, I think this should be ready to get merged

Reply:
Works for me. Thanks